### PR TITLE
Rebuild hungarian assigner

### PR DESCRIPTION
**Pull request by kellenf**
add focal_loss in hungarian assigner and rebuild its structure

### CODE REVIEW

File: mmdet/core/bbox/assigners/hungarian_assigner.py [hunk 1]
```diff
@@ -38,17 +38,16 @@ class HungarianAssigner(BaseAssigner):
...
-        self.iou_weight = iou_weight
-        self.iou_calculator = build_iou_calculator(iou_calculator)
+                 cls_cost=dict(type='ClsSoftmaxCost', weight=1.),
+                 reg_cost=dict(type='BBoxL1Cost', weight=1.0),
+                 iou_cost=dict(
+                     type='IoUBasedCost',
+                     iou_mode='giou',
+                     iou_calculator=dict(type='BboxOverlaps2D'),
+                     weight=1.0)):
+        self.cls_cost_func = build_match_cost(cls_cost)
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/hungarian_assigner.py:48]: We can skip cls_cost_func and simply use cls_cost, reg_cost, iou_cost

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    def __call__(self, cls_pred, gt_labels):
+        """
+        Args:
+            cls_pred (Tensor): Predicted classification logits, shape
+                [num_query, num_class].
+            gt_labels (Tensor): Label of `gt_bboxes`, shape (num_gt,).
+        """
+        cls_pred = cls_pred.sigmoid()
+        neg_cost = -(1 - cls_pred + self.eps).log() * (
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/match_cost.py:48]: Can we simply use the API in FocalLoss here?

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            1 - self.alpha) * cls_pred.pow(self.gamma)
+        pos_cost = -(cls_pred + self.eps).log() * self.alpha * (
+            1 - cls_pred).pow(self.gamma)
+        cls_cost = pos_cost[:, gt_labels] - neg_cost[:, gt_labels]
+        return cls_cost * self.weight
+
+
+@MATCH_COST.register_module()
+class ClsSoftmaxCost(object):
+
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/match_cost.py:58]: The docstring for class is necessary

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        self.alpha = alpha
+        self.gamma = gamma
+        self.eps = eps
+
+    def __call__(self, cls_pred, gt_labels):
+        """
+        Args:
+            cls_pred (Tensor): Predicted classification logits, shape
+                [num_query, num_class].
+            gt_labels (Tensor): Label of `gt_bboxes`, shape (num_gt,).
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/match_cost.py:45]: Docstring includes `Returns`

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        # NLL is used, we approximate it in 1 - cls_score[gt_label].
+        # The 1 is a constant that doesn't change the matching,
+        # so it can be ommitted.
+        cls_score = cls_pred.softmax(-1)
+        cls_cost = -cls_score[:, gt_labels]
+        return cls_cost * self.weight
+
+
+@MATCH_COST.register_module()
+class IoUBasedCost(object):
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/match_cost.py:79]: IoUCost

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+@MATCH_COST.register_module()
+class IoUCost(object):
+    """IoUCost.
+
+     Args:
+         iou_calculator (dict): cfg such as dict(type='BboxOverlaps2D')
+         iou_mode (str): iou mode
+         weight (int | float): loss weight
+     Examples:
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/match_cost.py:144]: Empty line.

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+         >>> import torch
+         >>> self = IoUCost()
+         >>> bboxes = torch.FloatTensor([[1,1, 2, 2], [2, 2, 3, 4]])
+         >>> gt_bboxes = torch.FloatTensor([[0, 0, 2, 4], [1, 2, 3, 4]])
+         >>> self(bboxes, gt_bboxes)
+         tensor([[-0.1250,  0.1667],
+                [ 0.1667, -0.5000]])
+     """
+    def __init__(self,
+                 iou_calculator=dict(type='BboxOverlaps2D'),
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/match_cost.py:155]: At this time, we may not need to use the encapsulated `iou_calculator` but directly use the function giou and iou.

File: tests/test_models/test_heads.py [hunk 8]
```diff
@@ -1243,11 +1243,13 @@ def test_transformer_head_loss():
     train_cfg = dict(
         assigner=dict(
             type='HungarianAssigner',
-            cls_weight=1.,
-            bbox_weight=5.,
-            iou_weight=2.,
-            iou_calculator=dict(type='BboxOverlaps2D'),
-            iou_mode='giou'))
+            cls_cost=dict(type='ClsSoftmaxCost', weight=1.0),
```
Comment (ZwwWayne) [inline tests/test_models/test_heads.py:1246]: ClsSoftmaxCost -> ClassificationCost

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        """
+        gt_bboxes_normalized = gt_bboxes / factor
+        # MODYFY: xyxy p1-norm in sparse-RCNN CODE which need to review
+        bbox_cost = torch.cdist(
+            bbox_pred, bbox_xyxy_to_cxcywh(gt_bboxes_normalized), p=1)
+        return bbox_cost * self.weight
+
+
+@MATCH_COST.register_module()
+class ClsFocalCost(object):
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/match_cost.py:49]: FocalLossCost

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            bbox_pred (Tensor): Predicted boxes with normalized coordinates
+                (cx, cy, w, h), which are all in range [0, 1]. Shape
+                [num_query, 4].
+            gt_bboxes (Tensor): Ground truth boxes with unnormalized
+                coordinates (x1, y1, x2, y2). Shape [num_gt, 4].
+            factor (Tensor): Image size(whwh) tensor. Shape [4].
+        """
+        gt_bboxes_normalized = gt_bboxes / factor
+        # MODYFY: xyxy p1-norm in sparse-RCNN CODE which need to review
+        bbox_cost = torch.cdist(
```
[the remark this answers is not in the document]
Comment (jshilong) [reply mmdet/core/bbox/match_costers/match_cost.py:43]: If you want to keep the `scale_factor`, maybe it is better to add the default value [1,1,1,1], and also normalize `bbox_pred` at the same time, or just delete it,  because  we can normalize `bbox_pred` and `gt_bboxes` with `whwh` outside of assign when we need(such as in sparse rcnn), because the normalization does not affect the value of ` iou` or `giou` loss

File: mmdet/core/bbox/match_costers/match_cost.py (hunk 9 above)
Comment (kellenf) [reply mmdet/core/bbox/match_costers/match_cost.py]: updated

File: mmdet/core/bbox/match_costers/match_cost.py (hunk 6 above)
Comment (kellenf) [reply mmdet/core/bbox/match_costers/match_cost.py:149]: updated

File: mmdet/core/bbox/match_costers/match_cost.py (hunk 7 above)
Comment (kellenf) [reply mmdet/core/bbox/match_costers/match_cost.py]: updated

File: tests/test_models/test_heads.py (hunk 8 above)
Comment (kellenf) [reply tests/test_models/test_heads.py]: updated

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+@MATCH_COST.register_module()
+class FocalLossCost(object):
+    """FocalLossCost.
+
+     Args:
+         weight (int | float): loss_weight
+         alpha (int | float): focal_loss alpha
+         gamma (int | float): focal_loss gamma
+         eps (float): default 1e-12
+     Examples:
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/match_cost.py:60]: 1. An empty line above examples.
2. eps (float, optional), gamma (int | float, optional)

File: mmdet/core/bbox/match_costers/match_cost.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+from .builder import MATCH_COST
+
+
+@MATCH_COST.register_module()
+class BBoxL1Cost(object):
+    """BBoxL1Cost.
+
+     Args:
+         weight (int | float): loss_weight
+     Examples:
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/match_cost.py:14]: 1. An empty line above Examples.
2. Update args accordingly.

File: mmdet/core/bbox/assigners/hungarian_assigner.py [hunk 13]
```diff
@@ -38,17 +38,15 @@ class HungarianAssigner(BaseAssigner):
...
-                 iou_weight=1.,
-                 iou_calculator=dict(type='BboxOverlaps2D'),
-                 iou_mode='giou'):
-        # defaultly giou cost is used in the official DETR repo.
-        self.iou_mode = iou_mode
-        self.cls_weight = cls_weight
-        self.bbox_weight = bbox_weight
-        self.iou_weight = iou_weight
-        self.iou_calculator = build_iou_calculator(iou_calculator)
+                 cls_cost=dict(type='ClassificationCost', weight=1.),
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/hungarian_assigner.py:41]: Update docstrings accordingly.

File: mmdet/core/bbox/match_costers/match_cost.py (hunk 11 above)
Comment (kellenf) [reply mmdet/core/bbox/match_costers/match_cost.py:62]: updated

File: mmdet/core/bbox/assigners/hungarian_assigner.py (hunk 13 above)
Comment (kellenf) [reply mmdet/core/bbox/assigners/hungarian_assigner.py:41]: updated

File: mmdet/core/bbox/match_costers/match_cost.py (hunk 12 above)
Comment (kellenf) [reply mmdet/core/bbox/match_costers/match_cost.py:16]: updated

**Review by kellenf — COMMENTED** [review]
reviewed

File: mmdet/core/bbox/match_costers/builder.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
+from mmcv.utils import Registry, build_from_cfg
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costers/builder.py:1]: Rename the directory, match_costers -> match_costs

Comment (kellenf) [reply mmdet/core/bbox/match_costers/builder.py]: updated

**Review by kellenf — COMMENTED** [review]
renamed match_costers -> match_costs

**Comment by ZwwWayne** [comment]
Please merge master to resolve conflicts.

**Comment by kellenf** [comment]
merged master

**Review by kellenf — COMMENTED** [review]
reviewed

**Review by ZwwWayne — APPROVED** [review]
